File: packages/core/src/session.ts
import type { CameraSession } from "./types";

/** Decode a JWT payload without verification (client-side only). */
export function decodeJwtPayload(token: string): Record<string, unknown> {
  const payloadB64 = token.split(".")[1] ?? "";
  const base64 = payloadB64.replace(/-/g, "+").replace(/_/g, "/");
  const padded  = base64 + "=".repeat((4 - (base64.length % 4)) % 4);
  return JSON.parse(atob(padded)) as Record<string, unknown>;
}

export function isTokenExpired(token: string): boolean {
  try {
    const payload = decodeJwtPayload(token);
    if (typeof payload.exp !== "number") return false;
    return Date.now() / 1000 > payload.exp;
  } catch {
    return true;
  }
}

export function sessionFromToken(token: string): CameraSession {
  const p = decodeJwtPayload(token) as {
    projectId: string;
    cameraId: string;
    cameraName: string;
    type: "interactive" | "passive";
  };
  return { token, projectId: p.projectId, cameraId: p.cameraId, cameraName: p.cameraName, type: p.type };
}

File: packages/flutter/lib/src/session.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import 'types.dart';

const _kSessionKey = 'meeeetup_cam_session';

/// Decode a JWT payload without verification (client-side only).
Map<String, dynamic> decodeJwtPayload(String token) {
  final parts = token.split('.');
  if (parts.length < 2) throw FormatException('Invalid JWT');
  var payload = parts[1]
      .replaceAll('-', '+')
      .replaceAll('_', '/');
  // Add padding
  while (payload.length % 4 != 0) {
    payload += '=';
  }
  final decoded = utf8.decode(base64.decode(payload));
  return json.decode(decoded) as Map<String, dynamic>;
}

bool isTokenExpired(String token) {
  try {
    final payload = decodeJwtPayload(token);
    final exp = payload['exp'];
    if (exp == null) return false;
    return DateTime.now().millisecondsSinceEpoch / 1000 > (exp as num);
  } catch (_) {
    return true;
  }
}

CameraSession sessionFromToken(String token) {
  final p = decodeJwtPayload(token);
  return CameraSession(
    token:      token,
    projectId:  p['projectId']  as String,
    cameraId:   p['cameraId']   as String,
    cameraName: p['cameraName'] as String,
    type:       p['type']       as String,
  );
}

Future<CameraSession?> loadSession() async {
  final prefs = await SharedPreferences.getInstance();
  final raw   = prefs.getString(_kSessionKey);
  if (raw == null) return null;
  try {
    final map     = json.decode(raw) as Map<String, dynamic>;
    final session = CameraSession.fromJson(map);
    if (isTokenExpired(session.token)) {
      await clearSession();
      return null;
    }
    return session;
  } catch (_) {
    return null;
  }
}

Future<void> saveSession(CameraSession session) async {
  final prefs = await SharedPreferences.getInstance();
  await prefs.setString(_kSessionKey, json.encode(session.toJson()));
}

Future<void> clearSession() async {
  final prefs = await SharedPreferences.getInstance();
  await prefs.remove(_kSessionKey);
}
